### ISSUE TITLE
zstd message + dictionary compression

### ISSUE THREAD

**Issue by mmeeks** [CollaboraOnline/online#6373]
We worked hard to better compress tile data and deltas which have quite a bit of internal redundancy which saves lots of space.

Unfortunately - since we want to distribute tiles compressed once to many different clients, and we don't want to re-compress the tiles constantly - we can't exploit compression over time / between multiple tiles / deltas (though this is done temporally with deltas of course).

However - [zstd](http://facebook.github.io/zstd/) has a dictionary mode, which it claims makes the (already fast) compression ~4x faster or so ;-) which would be appreciated too. So:

- we should find (or tweak) a JS implementation of zstd de-compression so that it will accept a dictionary to base off =)
- we should (at build time I guess?) construct a dictionary from some plausible test data set of documents with text in, as well as a set of deltas, but probably particularly also the many initialization messages that we tend to send at startup: font-lists etc. as JSON.
-  we should save the dictionary and serve it to the JS using coolwsd: so it can be cached on the client for this coolwsd/hash - once and be used to accelerate every subsequent start / use of the protocol.
-  we should use ZSTD_compress_using_CDict - or some equivalent: "When compressing multiple messages or blocks using the same dictionary, it's recommended to digest the dictionary only once, since it's a costly operation."
- we should look at types of messages where the message is not processed by coolwsd but emitted straight from the Kit <-> JS using coolwsd as a pure transport [ eg. modified-state, invalidations and many other things are quite critical for coolwsd to watch to account for things properly ]. Anyhow @eszkadev 's JSDialog dialog JSON would be a great candidate for dictionary compression, along with lists of font names I guess. We compress only these messages (one at a time) vs. this dictionary and push them through our pipeline, perhaps as binary WebSocket messages.
- we de-compress them inside the Socket code at the other end; cf. src/layer/tile/CanvasTileLayer.js: var allDeltas = window.fzstd.decompress(rawDelta);
-  we use some new shiny JS implementation that supports dictionary primed de-compression.

Thanks !

**Comment by jazevedo-coll** [comment]
@mmeeks is this ticket still relevant?

**Comment by mmeeks** [comment]
Yes - still relevant; if we got it right, could reduce quite some size in our protocol - but we'd need work on the zstd JS piece.

**Comment by jazevedo-coll** [comment]
ok, then adding 25.04 tag here.
